Don't require dev-dependencies when not needed in certain cases

infinity0 opened the pull request:
Specifically, when running `cargo install` and add a flag for this behaviour in `cargo build`.

This closes #4988.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @alexcrichton (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

alexcrichton commented on src/bin/build.rs:
```diff
@@ -63,6 +64,7 @@ Options:
     --features FEATURES          Space-separated list of features to also build
     --all-features               Build all available features
     --no-default-features        Do not build the `default` feature
+    --avoid-dev-deps             Avoid installing dev-dependencies if possible
```
Could this start out as a `-Z` flag to be unstable by default?

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -413,9 +419,21 @@ impl<'a> CompileFilter<'a> {
+                TargetKind::Bin => true,
+                TargetKind::Lib(..) => true,
+                _ => false,
+            },
```
How come this needed changing?

infinity0 replied:
It's not necessary and could be dropped from this PR. But without it, the current situation means I have to give extra flags to `cargo build`, it contradicts the documentation, and it also means the `all-targets` flag is redundant.

alexcrichton replied:
Hm sorry I missed this earlier, but I'm not sure I quite understand this. Could this perhaps be left out for a future PR?

alexcrichton replied:
(or maybe a test could be added?)

infinity0 replied:
This actually opened up a whole new can of worms, so I've reverted it from this PR and opened up a new issue #5134 which I'll file a PR for as soon as this PR is merged (I have a fix, but it builds on top of this PR).

alexcrichton commented on src/cargo/ops/cargo_install.rs:
```diff
@@ -175,7 +175,11 @@ fn install_one(root: &Filesystem,
 
     let ws = match overidden_target_dir {
         Some(dir) => Workspace::ephemeral(pkg, config, Some(dir), false)?,
```
Shouldn't the optional deps flag be set for this as well?

infinity0 replied:
ephemeral() already sets that internally

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -226,12 +226,18 @@ pub fn compile_ws<'a>(ws: &Workspace<'a>,
-                                            &specs)?;
+    let features = Method::split_features(features);
+    let method = Method::Required {
+        dev_deps: ws.require_optional_deps() || filter.need_dev_deps(),
```
Perhaps the name `optional_deps` could change since it's affecting dev-dependencies?

Additionally, I think that we'd want this flag to affect target-specific dependencies, right? If this is a sort of "one shot" compilation there should also be no need to resolve and require windows-specific dependencies when on Linux, right?

infinity0 replied:
Would you mind if we leave that as a TODO for the future? What you say makes sense but it is not actually needed in Debian right now, because we union over all targets when translating dependencies - so e.g. winapi etc are all pulled in regardless of what target you're building for. This means we have to have less special-cases in the packaging, and also supports cross-compilation later.

If I understood correctly, on Fedora they are patching out windows dependencies so they don't appear in Cargo.toml, so this functionality wouldn't be needed either. (Perhaps @ignatenkobrain can confirm).

alexcrichton replied:
Ok sure yeah, I think a TODO should suffice!

infinity0 replied:
I'm unsure if it's appropriate to mention "targets" in the sense of "platform" here on this line. The term "target" in the context of CompileFilter and `core::Target` and `cargo build` options, seems rather to mean "lib/bin/examples/tests" etc and not platform.

OTOH the functionality you mention does seem to be missing - if I add `[target."XXXX".dependencies] YYY = "999"` to Cargo.toml then `cargo install` still fails with this PR, despite the fact that "XXXX" does not match the current platform. I'm just unsure of the appropriate place to add this TODO, please advise. Perhaps on `struct Method`?

alexcrichton replied:
Hm ok, perhaps an issue can be filed? I think the `-Z avoid-dev-deps` flag added here may in the long run want to be something like `--minimal-cargo-lock` which prunes all non-relevant dependencies like platform-specific dependencies that don't apply, dev-deps if you're not building tests, etc. In that sense I think the feature/TODO here is slightly broader than a comment in the code. Would you be ok filing that?

infinity0 replied:
Filed as #5133

infinity0 commented:
Thanks for the comments, I'll soon find some time to write tests and switch to `-Z` for the new flag.

infinity0 commented:
Hey, I've added some tests and changed `avoid-dev-deps` to be a `-Z` unstable flag.

I wasn't quite sure what to rename `require_optional_deps` to, and it's a public struct so switching would carry some extra cost. I did add a comment to it, describing the new behaviour. I could change it if you want though, like `require_extra_deps` or `require_unneeded_deps`, let me know which you prefer.

bors commented:
:umbrella: The latest upstream changes (presumably #5063) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
Thanks for the update, tests look good to me! It's also ok to not worry much about naming, despite being public we bump the major version for Cargo regularly so it's not too critical to think too hard about APIs

infinity0 commented:
I've resolved the merge conflict and dealt with all the remaining issues I think, please take a look and merge if good.

alexcrichton commented:
Ah looks like some tests may be failing?

infinity0 commented:
Turns out the tests were also dependent on the fix for #5134, I've pushed a commit to work around it. Should be all good now.

alexcrichton commented:
@bors: r+

bors commented:
:pushpin: Commit d46db71 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton d46db71b3ff17dfc0f4be6308c8b94613d65a572 -->

bors commented:
:hourglass: Testing commit d46db71b3ff17dfc0f4be6308c8b94613d65a572 with merge 879f48324d52c78322b6ad7513999b2a651c5023...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/cargo/build/1.0.3835), [status-travis](https://travis-ci.org/rust-lang/cargo/builds/350376287?utm_source=github_status&utm_medium=notification)
Approved by: alexcrichton
Pushing 879f48324d52c78322b6ad7513999b2a651c5023 to master...
